### PR TITLE
Resolv code hash

### PR DESCRIPTION
**Pull request by MathieuDutSik**
This branch resolves problem with the copy paste of code.
As a consequence, there is now a new program 'setup_contract'
for setting up the contract.
Both the side_chain_client and the side_chain_manager uses a file
named 'contract_address.json' that contains:
contract_address / code_hash / transaction_hash / creation_block
which are of course checked for consistency when loading.
The PR introduces some changes on the docker side of things.
There is an infinite loop checking for the file to be present when
loading it.
The PR introduces various cleanups elsewhere.

### CODE REVIEW

File: src/alacris_lib/operator_contract.ml [hunk 1]
```diff
@@ -24,58 +27,166 @@ let topic_of_hash (hash : Digest.t) : Bytes.t option =
...
+  ; creation_block : int
+  } [@@deriving of_yojson]
+
+
+type contract_address_config =
+  { contract_address : Address.t
+  ; code_hash : Digest.t
+  ; creation_hash : Digest.t
+  ; creation_block : Revision.t
+  }
```
Comment (fare) [inline src/alacris_lib/operator_contract.ml]: Shouldn't the two be the very same type, using the Address and Digest and Revision yojson support to read the string properly?

File: src/alacris_lib/operator_contract.mli [hunk 2]
```diff
@@ -13,23 +13,23 @@ val topic_of_amount:   TokenAmount.t -> Bytes.t option
...
+  ; code_hash : Digest.t
+  ; creation_hash : Digest.t
+  ; creation_block : Revision.t
+  }
 
 
-val set_contract_address : Address.t -> unit
-(** set the address of the contract on Ethereum
-    TODO: use a fixed address, obviating this call (?) *)
+val retrieve_contract_address_quadruple : Digest.t -> contract_address_config Lwt_exn.t
```
Comment (fare) [inline src/alacris_lib/operator_contract.mli]: Replace `contract_address_quadruple` and `contract_address_config` by just `contract_config` ?

File: src/legilogic_ethereum/ethereum_user.ml [hunk 3]
```diff
@@ -314,6 +325,29 @@ let confirmed_or_known_issue : Address.t -> (Digest.t, TransactionReceipt.t) Lwt
 
 exception Replacement_transaction_underpriced
 
+
+let limited_retry : int -> 'a -> ('a -> 'b option Lwt_exn.t) -> 'b option Lwt_exn.t =
```
Comment (fare) [inline src/legilogic_ethereum/ethereum_user.ml]: Why not use the usual retry function?

File: src/legilogic_ethereum/ethereum_user.ml [hunk 4]
```diff
@@ -368,12 +405,49 @@ let send_and_confirm_transaction : (Transaction.t * SignedTransaction.t, Transac
...
+           fail Still_pending))
     >>= check_receipt_sufficiently_confirmed
 
+
+
+
+
+
+
+let infinite_loop_original : (Transaction.t * SignedTransaction.t) -> TransactionReceipt.t Lwt_exn.t =
```
Comment (fare) [inline src/legilogic_ethereum/ethereum_user.ml]: What is this function for? It needs a better name than that.

File: src/legilogic_ethereum/ethereum_user.ml (hunk 3 above)
Comment (MathieuDutSik) [reply src/legilogic_ethereum/ethereum_user.ml]: eliminated

File: src/legilogic_ethereum/ethereum_user.ml (hunk 4 above)
Comment (MathieuDutSik) [reply src/legilogic_ethereum/ethereum_user.ml]: Eliminated and put back the code as before.

File: src/alacris_lib/operator_contract.mli (hunk 2 above)
Comment (MathieuDutSik) [reply src/alacris_lib/operator_contract.mli]: Renaming has been done

File: src/alacris_lib/operator_contract.ml (hunk 1 above)
Comment (MathieuDutSik) [reply src/alacris_lib/operator_contract.ml]: This is for sure the natural solution.
However, the problem is that the Digest.t does not have a "of_yojson" implementation.
I tried to implement the Digest.of_yojson but without success.

File: src/alacris_contract/setup_contract.ml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+let setup_contract_log = false
+
+
+let _ =
+  Lwt_exn.run
+    (fun () ->
+      if setup_contract_log then
+        Logging.log "Before the Db.open_connection";
+      of_lwt Db.open_connection "setup_contract_db"
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply src/alacris_contract/setup_contract.ml]: Ok, I implemented this and it works.

File: src/legilogic_ethereum/ethereum_user.ml [hunk 6]
```diff
@@ -611,7 +647,7 @@ let make_pre_transaction ~sender operation ?gas_limit ~value : PreTransaction.t
   if ethereum_user_log then
     Logging.log "ETHUSR: make_pre_transaction gas_limit=%s value=%s" (TokenAmount.to_string gas_limit) (TokenAmount.to_string value);
   (* TODO: The multiplication by 2 is a hack that needs to be addressed *)
-  let gas_limit_n_fold = (TokenAmount.mul (TokenAmount.of_int 2) gas_limit) in
+  let gas_limit_n_fold = (TokenAmount.mul (TokenAmount.of_int 3) gas_limit) in
```
Comment (fare) [inline src/legilogic_ethereum/ethereum_user.ml]: Why do we need increase the limit that much, already? I thought you had identified and fixed the problem with gas calculation.

Note that metamask seems to add 50% to the gas estimate and be happy with it. x3 seems a bit excessive.

File: src/alacris_lib/operator_contract.ml [hunk 7]
```diff
@@ -24,58 +27,166 @@ let topic_of_hash (hash : Digest.t) : Bytes.t option =
...
+  | Some receipt ->
+     let contract_address = Option.get receipt.contract_address in
+     let contract_block_number = receipt.block_number in
+     Ethereum_json_rpc.eth_get_transaction_by_hash transaction_hash
+     >>= fun trans ->
+     let code_hash = Digesting.digest_of_string (Bytes.to_string trans.input) in
+     return {contract_address; code_hash; creation_hash=transaction_hash; creation_block=contract_block_number}
+
+
+let get_contract_address_general : unit -> Address.t Lwt_exn.t =
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply src/alacris_lib/operator_contract.ml]: It is not exported. I renamed it as **_internal**.

File: src/alacris_lib/side_chain_action.ml [hunk 8]
```diff
@@ -11,50 +10,23 @@ open Side_chain_user
...
-  >>= fun (contract_address, code_hash, creation_hash, creation_block) ->
-  Logging.log "contract_address=%s" (Address.to_0x contract_address);
-  Logging.log "code_hash=%s" (Digest.to_0x code_hash);
-  Logging.log "creation_hash=%s" (Digest.to_0x creation_hash);
-  Logging.log "creation_block=%s" (Revision.to_string creation_block);
-  Logging.log "E N T R I E S to put in the side_chain_client_config.json file";
-  Logging.log "  \"contract_address\": \"%s\",\n  \"code_hash\": \"%s\",\n  \"creation_hash\": \"%s\",\n  \"creation_block\": %s," (Address.to_0x contract_address) (Digest.to_0x code_hash) (Digest.to_0x creation_hash) (Revision.to_string creation_block);
-  Address.to_0x contract_address
+  Operator_contract.retrieve_contract_config transaction_hash
+  >>= fun e_quad ->
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply src/alacris_lib/side_chain_action.ml]: Ok, renamed.

File: src/legilogic_ethereum/ethereum_user.ml (hunk 6 above)
Comment (MathieuDutSik) [reply src/legilogic_ethereum/ethereum_user.ml]: Ok, reverted to 2.

File: src/alacris_lib/side_chain_action.ml [hunk 9]
```diff
@@ -101,25 +68,21 @@ module Test = struct
...
-        Operator_contract.set_contract_address contract_address;
-
         (* TODO consolidate integration tests into single entry point with
          * shared initialization phase rather than leaving them scattered about
          * the repo. One problem with the present setup is we cannot shut down
          * the following reactor once flipping it on, meaning we're likely to
          * encounter subtle time-dependent bugs in future tests (until we
          * reorganize) *)
-        State_update.start_state_update_operator () >>= fun _ ->
+        State_update.start_state_update_operator ()
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply src/alacris_lib/side_chain_action.ml]: Ok, renamed to **_ademon**

File: src/alacris_lib/operator_contract.ml [hunk 10]
```diff
@@ -24,58 +27,166 @@ let topic_of_hash (hash : Digest.t) : Bytes.t option =
...
+
+type contract_address_config =
+  { contract_address : Address.t
+  ; code_hash : Digest.t
+  ; creation_hash : Digest.t
+  ; creation_block : Revision.t
+  }
+
+
+
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply src/alacris_lib/operator_contract.ml]: Yes, this is what I did.

File: src/alacris_lib/operator_contract.ml [hunk 11]
```diff
@@ -24,58 +27,166 @@ let topic_of_hash (hash : Digest.t) : Bytes.t option =
...
 
 
+
+type contract_address_config_input =
+  { contract_address : string
+  ; code_hash : string
+  ; creation_hash : string
+  ; creation_block : int
+  } [@@deriving of_yojson]
+
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply src/alacris_lib/operator_contract.ml]: Yes, this is what I did.

**Pull request merged**